# Loan Data from Prosper Exploration
## by Jeff Mitchell

## Preliminary Wrangling

This dataset contains nearly 114,000 loans from Prosper.

In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

> Load in your dataset and describe its properties through the questions below.
Try and motivate your exploration goals through this section.

In [3]:
loans = pd.read_csv('prosperLoanData.csv')

In [19]:
# high-level overview of data shape and composition
print(loans.shape)

(113937, 81)


In [6]:
# descriptive statistics for numeric variables
print(loans.describe())

       ListingNumber           Term    BorrowerAPR   BorrowerRate  \
count   1.139370e+05  113937.000000  113912.000000  113937.000000   
mean    6.278857e+05      40.830248       0.218828       0.192764   
std     3.280762e+05      10.436212       0.080364       0.074818   
min     4.000000e+00      12.000000       0.006530       0.000000   
25%     4.009190e+05      36.000000       0.156290       0.134000   
50%     6.005540e+05      36.000000       0.209760       0.184000   
75%     8.926340e+05      36.000000       0.283810       0.250000   
max     1.255725e+06      60.000000       0.512290       0.497500   

         LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
count  113937.000000             84853.000000   84853.000000     84853.000000   
mean        0.182701                 0.168661       0.080306         0.096068   
std         0.074516                 0.068467       0.046764         0.030403   
min        -0.010000                -0.182700       0.

In [18]:
loans['LoanStatus'].value_counts()

Current                   56576
Completed                 38074
Chargedoff                11992
Defaulted                  5018
Past Due (1-15 days)        806
Past Due (31-60 days)       363
Past Due (61-90 days)       313
Past Due (91-120 days)      304
Past Due (16-30 days)       265
FinalPaymentInProgress      205
Past Due (>120 days)         16
Cancelled                     5
Name: LoanStatus, dtype: int64

In [32]:
# Convert the LoanStatus column to categorical types
loans['LoanStatus'] = loans['LoanStatus'].astype('category')

In [17]:
dupes = loans[loans['LoanNumber'].duplicated() == True]
dupes.sort_values('LoanNumber')

ListingKey  ListingNumber            ListingCreationDate  \
101337  F7313585219017186FA3C46         859848  2013-08-04 19:00:15.727000000   
26357   2F1D3586569876887E1B09D         875651  2013-08-21 05:36:58.137000000   
50761   CDFB3588980067860F9C7F3         893227  2013-09-12 16:56:42.890000000   
66322   C17535874571157867FA321         885990  2013-09-03 08:52:26.490000000   
64633   5CEB35892993375750F2B06         894308  2013-09-11 15:57:35.400000000   
102234  64E33587560804418EAAE9F         887683  2013-09-04 09:01:14.540000000   
10401   68153589168873924D3A78D         898052  2013-09-12 12:53:22.680000000   
58686   4BC835901005727268E6551         905592  2013-09-17 17:21:30.453000000   
86717   958D358643285462008F535         864112  2013-08-07 20:23:57.063000000   
96475   B6A835897654847592040D8         900306  2013-09-11 16:04:18.790000000   
73145   8BBA3585140881118FBD815         858364  2013-08-01 18:51:46.200000000   
58914   A4D93589338268643F7CA07         899828  2013-09-12 19:05:08.637000000   
111890  2AEA358216796526862A4CF         827053  2013-07-02 14:09:36.273000000   
87265   45C2357870288848093FB87         786407  2013-05-22 19:42:23.417000000   
103559  6A833587133492369F9D00F         880017  2013-08-26 10:53:26.637000000   
32905   C671358665442666939A17F         879829  2013-08-24 17:10:40.913000000   
66242   37FF3586674023430A91562         866202  2013-08-11 05:35:48.853000000   
33649   C9FF35684194774626F5F95         701942  2013-01-20 21:07:13.620000000   
79482   3D6A35900355229258C47D7         910395  2013-09-19 09:34:07.627000000   
80961   CCDD3590323665222C1BD9E         918397  2013-09-21 19:15:24.510000000   
110390  915B3587469694068D2084D         884862  2013-09-01 01:44:51.940000000   
49392   A33C35896968384756D05FE         899615  2013-09-12 16:56:41.283000000   
99817   D729358878305631215B1D9         901899  2013-09-13 19:03:52.513000000   
109074  22813584531824926A3F46C         852758  2013-07-27 00:09:59.597000000   
70554   07F235851119053934BDC2B         861876  2013-08-06 14:14:55.543000000   
110185  262B359008715750187E76A         905260  2013-09-17 17:16:21.417000000   
59216   D7F53583080367563F57548         836614  2013-07-11 17:13:52.567000000   
84055   59113589279069224B256B2         910144  2013-09-18 16:47:22.237000000   
95303   9633358913954552068C392         917584  2013-09-21 13:47:47.473000000   
87606   45453589039926473267B8B         893972  2013-09-11 14:25:29.697000000   
...                         ...            ...                            ...   
35342   185F360439569801000C899        1237178  2014-03-06 09:40:39.300000000   
100708  580B3604039442380CCA906        1212311  2014-02-27 12:51:25.090000000   
2539    2D2635984503681189056B4        1097584  2014-01-11 07:50:08.037000000   
100747  DE0C360292962505299EE67        1224533  2014-03-04 10:07:02.010000000   
61571   A2973600547802538F3E028        1146677  2014-01-18 12:03:24.550000000   
31659   D1E03603106629791E8F273        1200566  2014-02-20 09:36:41.433000000   
71776   C37036030063888627629F4        1220787  2014-02-18 14:30:09.347000000   
58037   B9183604327597346AE6624        1230428  2014-03-05 07:31:12.570000000   
35990   1987359827858610575623F        1114520  2014-01-02 18:34:42.030000000   
100901  DE8A3598254689530D74484        1115681  2014-01-04 06:03:55.070000000   
107162  C58F3599910678266283637        1159061  2014-01-24 13:12:01.193000000   
100281  F4E235969724800397FFB3B        1081388  2013-12-15 18:41:23.297000000   
51906   924535997868157428B74CA        1136469  2014-01-14 05:21:35.957000000   
37563   103B36002725491878567E3        1113064  2014-01-16 05:51:49.257000000   
97273   26C835968174004476E551B        1017064  2013-11-29 11:05:25.743000000   
91557   26C835968174004476E551B        1017064  2013-11-29 11:05:25.743000000   
90365   4160360046339713343541D        1125268  2014-01-21 13:22:06.683000000   
35093   31B8360054573338082BC74        1138904  2014

### What is the structure of your dataset?

There are 113,937 loans with 81 features for each loan. This is a lot of features and some are only relevant to certain periods of time (e.g. pre-2009, post-July 2009). These features include details such as Term, LoanStatus, BorrowerRate, various scores and ratings, demographic features such as Occupation, EmploymentStatus, Home ownership, various credit scores and totals, income range and details specific t the loan.

### What is/are the main feature(s) of interest in your dataset?

I am most interested in finding out what features have the greatest impact on Loan Outcome Status.

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

There are a number of features that I feel to be of interest in investigating Loan Outcome Status. These include 

There are a lot of columns in the data which makes useful analysis difficult. I will start by reducing the dataframe to only the columns that are of interest to me for investigating the Loan Status Outcome.

In [21]:
# Reduce the number of columns to just those that may be of interest
columns = ['Term', 'LoanStatus', 'BorrowerAPR', 'BorrowerRate', 'ListingCategory (numeric)', 'EmploymentStatus', 
          'IsBorrowerHomeowner', 'CreditScoreRangeLower', 'CreditScoreRangeUpper', 'CurrentCreditLines',
           'TotalCreditLinespast7years', 'OpenRevolvingAccounts', 'OpenRevolvingMonthlyPayment', 'CurrentDelinquencies',
          'AmountDelinquent', 'DelinquenciesLast7Years', 'RevolvingCreditBalance', 'BankcardUtilization',
           'DebtToIncomeRatio', 'StatedMonthlyIncome', 'LoanNumber', 'LoanOriginalAmount', 'LoanOriginationQuarter',
          'MemberKey', 'MonthlyLoanPayment']
loans = loans[columns]

Some of the column headings are confusing or difficult to work with so I will rename them.

In [22]:
loans = loans.rename(columns={'ListingCategory (numeric)':'ListingCategory','IsBorrowerHomeowner':'HomeOwner',
                              'TotalCreditLinespast7years':'TotalCreditLines',
                              'DelinquenciesLast7Years':'TotalDelinquencies'})

Term LoanStatus  BorrowerAPR  BorrowerRate  ListingCategory  \
0    36  Completed      0.16516        0.1580                0   
1    36    Current      0.12016        0.0920                2   
2    36  Completed      0.28269        0.2750                0   
3    36    Current      0.12528        0.0974               16   
4    36    Current      0.24614        0.2085                2   

  EmploymentStatus  HomeOwner  CreditScoreRangeLower  CreditScoreRangeUpper  \
0    Self-employed       True                  640.0                  659.0   
1         Employed      False                  680.0                  699.0   
2    Not available      False                  480.0                  499.0   
3         Employed       True                  800.0                  819.0   
4         Employed       True                  680.0                  699.0   

   CurrentCreditLines  ...  TotalDelinquencies  RevolvingCreditBalance  \
0                 5.0  ...                 4.0                     0.0   
1                14.0  ...                 0.0                  3989.0   
2                 NaN  ...                 0.0                     NaN   
3                 5.0  ...                14.0                  1444.0   
4                19.0  ...                 0.0                  6193.0   

   BankcardUtilization  DebtToIncomeRatio  StatedMonthlyIncome  LoanNumber  \
0                 0.00               0.17          3083.333333       19141   
1                 0.21               0.18          6125.000000      134815   
2                  NaN               0.06          2083.333333        6466   
3                 0.04               0.15          2875.000000       77296   
4                 0.81               0.26          9583.333333      102670   

   LoanOriginalAmount  LoanOriginationQuarter                MemberKey  \
0                9425                 Q3 2007  1F3E3376408759268057EDA   
1               10000                 Q1 2014  1D13370546739025387B2F4   
2                3001                 Q1 2007  5F7033715035555618FA612   
3               10000                 Q4 2012  9ADE356069835475068C6D2   
4               15000                 Q3 2013  36CE356043264555721F06C   

   MonthlyLoanPayment  
0              330.43  
1              318.93  
2              123.32  
3              321.45  
4              563.97  

[5 rows x 25 columns]

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.

> Make sure that, after every plot or related series of plots, that you
include a Markdown cell with comments about what you observed, and what
you plan on investigating next.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!